### PR TITLE
8 feature 8 save player stats feature

### PR DESCRIPTION
**Pull request by lilyxmeng**
Created LoadPlayer and SavePlayer classes, along with SaveLoadPlayerTest test unit to check various cases. This PR adds data persistence to the project (save information and load that information later). The file to which information is being saved and loaded is player.csv, which is currently under the use_cases package. 

Concerns I would like addressed for review:
- Should player.csv be under use_cases or a lower layer? (Adhering to CA)
- Are there any edge cases that I've missed in my tests?

### CODE REVIEW

File: src/main/java/entities/HumanPlayer.java [hunk 1]
```diff
@@ -42,4 +42,19 @@ public void incrementLosses() {
         this.losses ++;
     }
 
+    /**
+     * Get the number of wins of this HumanPlayer.
+     * @return An int containing the number of wins of this HumanPlayer.
+     */
+    public int getWins() {
```
Comment (nyan-ko) [inline src/main/java/entities/HumanPlayer.java:49]: this is technically unnecessary because of the getStats() method here, however i prefer distinct methods as opposed to putting everything into an array. we should decide which behaviour to stick with

File: src/main/java/use_cases/LoadPlayer.java
Comment (nyan-ko) [inline src/main/java/use_cases/LoadPlayer.java:1]: i feel like something as high level as a use case shouldnt be handling the low level details, which here would be reading a csv file. we should define an interface that supports loading and saving, then make an implementation of that interface that specifically handles csv reading and writing.

File: src/main/java/use_cases/LoadPlayer.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+public class LoadPlayer {
+    /**
+     * Loads a HumanPlayer's wins and losses from "src/main/java/use_cases/players.csv" given their name.
+     * @param name The name of the HumanPlayer.
+     * @return A string array [name, wins, losses] of the HumanPlayer, or null if the name is not found.
+     * @throws IOException If an input/output error occurs when reading from the file.
+     */
+    public static String[] loadPlayerFromFile(String name) throws IOException {
+        String file = "src/main/java/use_cases/players.csv";
```
Comment (nyan-ko) [inline src/main/java/use_cases/LoadPlayer.java:15]: just as an example:

```
public interface DataAccess {
    PlayerInformation load(String name, String file)  // here PlayerInformation would just be a data object of all the player stats
    void save(String name, String file)
}
```

with an example implementation:

```
public class CSVDatabase implements DataAccess {
  ...
  
  @Override
  public PlayerInformation load(String name, String file) {
    BufferedReader br = ...
    // your existing code in LoadPlayer
  }

  @Override
  public void save(String name, String file) {
    // your code in SavePlayer
  }
}
```
your existing usecases would then store a reference to a DataAccess interface and call its method, and we'd pass in a CSVDatabase class wherever we initialize them

the specifics can be decided later but i think this is a good guideline for keeping use cases separate from the databases.

File: src/main/java/use_cases/SavePlayer.java
Comment (nyan-ko) [inline src/main/java/use_cases/SavePlayer.java:1]: same comment for LoadPlayer

**Review by nyan-ko — CHANGES_REQUESTED** [review]

**Comment by nyan-ko** [comment]
player.csv should be at a lower level than use cases. following clean architecture guidelines, the csv file is a low level detail and should be in the outermost layer, however i think we should package it with the rest of the csv database files, for example, in a `database` folder

tests look comprehensive to me 👍

File: src/main/java/entities/HumanPlayer.java (hunk 1 above)
Comment (SMateen21) [reply src/main/java/entities/HumanPlayer.java:49]: This behaviour is better I believe.

File: src/main/java/use_cases/SavePlayer.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+public class SavePlayer {
+    /**
+     * If the HumanPlayer does not exist in "src/main/java/use_cases/players.csv", saves its name, wins, and losses.
+     * If the HumanPlayer exists in "src/main/java/use_cases/players.csv", do not save anything.
+     * @param player The HumanPlayer object.
+     * @return true if information was saved, false if nothing was saved.
+     * @throws IOException If an input/output error occurs when writing to the file.
+     */
+    public static boolean savePlayerToFile(HumanPlayer player) throws IOException {
```
Comment (SMateen21) [inline src/main/java/use_cases/SavePlayer.java:17]: How/where would we handle this exception? That wouldn't be included here but just a thought.

File: src/main/java/use_cases/SavePlayer.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package use_cases;
```
Comment (SMateen21) [inline src/main/java/use_cases/SavePlayer.java:1]: I'm not too sure about including this in the use_cases packages, it is too concerned with the way we are saving Players (the "how"). I'd go with Edric's suggestion here.

**Review by SMateen21 — COMMENTED** [review]
Looks good! There is a Clean Architecture violation that can be avoided, I'd go with Edric's suggestion there before deciding to merge.

File: src/main/java/entities/HumanPlayer.java (hunk 1 above)
Comment (lilyxmeng) [reply src/main/java/entities/HumanPlayer.java:49]: Okay I'll be removing getStats(). There are a few tests in HumanPlayerTest that use the getStats() method so I'll be fixing it in my following commits.

File: src/main/java/use_cases/SavePlayer.java (hunk 4 above)
Comment (lilyxmeng) [reply src/main/java/use_cases/SavePlayer.java:1]: I had to include the use_cases package here because SavePlayer is in the use_cases package.

File: src/main/java/use_cases/LoadPlayer.java (hunk 2 above)
Comment (lilyxmeng) [reply src/main/java/use_cases/LoadPlayer.java]: I've pushed new changes according to this suggestion. Let me know if I should make further edits.

File: src/main/java/use_cases/CSVDatabase.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
+package use_cases;
```
Comment (SMateen21) [inline src/main/java/use_cases/CSVDatabase.java:1]: I'd change this package to be something like "drivers" or "database" , something like that, but other than that all is good.

**Review by SMateen21 — APPROVED** [review]
Looks great! The DataAccess interface will make it easier for use case classes to load players and save them. Only thing would be changing the package which CSVDatabase is under though, as I don't think it should be part of use_cases.

File: src/main/java/use_cases/SavePlayer.java (hunk 3 above)
Comment (lilyxmeng) [reply src/main/java/use_cases/SavePlayer.java]: The method will throw an IOException if it has an error reading or accessing the file. I don't think this is likely to happen given our relatively simple use of saving/loading data, but I could change this to a catch statement and explicitly handle it although I'm not sure what we would do.

File: src/main/java/use_cases/CSVDatabase.java (hunk 5 above)
Comment (nyan-ko) [reply src/main/java/use_cases/CSVDatabase.java:1]: agreed

File: src/main/java/use_cases/DataAccess.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import java.io.IOException;
+
+public interface DataAccess {
+    /**
+     * Loads a HumanPlayer's wins and losses from "src/main/java/use_cases/players.csv" given their name.
+     * @param name The name of the HumanPlayer.
+     * @return A string array [name, wins, losses] of the HumanPlayer, or null if the name is not found.
+     * @throws IOException If an input/output error occurs when reading from the file.
+     */
+    String[] loadPlayer(String name) throws IOException;
```
Comment (nyan-ko) [inline src/main/java/use_cases/DataAccess.java:14]: i think a data class instead of a string array is better for this return value

File: src/main/java/use_cases/CSVDatabase.java (hunk 5 above)
Comment (lilyxmeng) [reply src/main/java/use_cases/CSVDatabase.java:1]: That makes sense, I'll make that change.

File: src/main/java/use_cases/DataAccess.java (hunk 6 above)
Comment (lilyxmeng) [reply src/main/java/use_cases/DataAccess.java]: Got it, created PlayerInformation class.

File: src/test/java/database/SaveLoadPlayerTest.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class SaveLoadPlayerTest {
+
+    private HumanPlayer p1;
+    private HumanPlayer p2;
+    private HumanPlayer p3;
+    
+    CSVDatabase database;
```
Comment (nyan-ko) [inline src/test/java/database/SaveLoadPlayerTest.java:20]: add access modifier

Comment (lilyxmeng) [reply src/test/java/database/SaveLoadPlayerTest.java]: Fixed

**Review by nyan-ko — APPROVED** [review]
LGTM!

File: src/main/java/database/CSVDatabase.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            String[] values = line.split(",");
+            if (values[0].equals(name)) {
+                return new PlayerInformation(values[0], Integer.parseInt(values[1]), Integer.parseInt(values[2]));
+            }
+        }
+        return null;
+    }
+
+    /**
+     * If the HumanPlayer does not exist in "src/main/java/use_cases/players.csv", saves its name, wins, and losses.
```
Comment (meiiph) [inline src/main/java/database/CSVDatabase.java:32]: save* (instead of saves)

File: src/main/java/database/DataAccess.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    /**
+     * Loads a HumanPlayer's wins and losses from "src/main/java/use_cases/players.csv" given their name.
+     * @param name The name of the HumanPlayer.
+     * @return A PlayerInformation object, or null if the name is not found.
+     * @throws IOException If an input/output error occurs when reading from the file.
+     */
+    PlayerInformation loadPlayer(String name) throws IOException;
+
+    /**
+     * If the HumanPlayer does not exist in "src/main/java/use_cases/players.csv", saves its name, wins, and losses.
```
Comment (meiiph) [inline src/main/java/database/DataAccess.java:18]: same comment as the one I just made, save* rather than saves :)

File: src/main/java/usecases/PlayerInformation.java
Comment (meiiph) [inline src/main/java/usecases/PlayerInformation.java:1]: Anywhere "this players" appears should be replaced with "this player's" in this file.

File: src/test/java/database/SaveLoadPlayerTest.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        assert p4 != null;
+        assertEquals(p1.getName(), p4.getName());
+        assertEquals(p1.getWins(), p4.getWins());
+        assertEquals(p1.getLosses(), p4.getLosses());
+
+        assertNull(p5);
+    }
+
+    /**
+     * Saves an existing player (should not save).
```
Comment (meiiph) [inline src/test/java/database/SaveLoadPlayerTest.java:130]: Is this supposed to save or not save the player? The wording is a bit confusing here.

**Review by meiiph — APPROVED** [review]
Good, just fix the Javadocs comments that I added. :)

**Review by SMateen21 — APPROVED** [review]
Everything looks good. I'd add some light Javadocs for the classes as well before merging but otherwise I approve.
